pygmt/tests/test_blockmean.py

```diff
@@ -0,0 +1,78 @@
+"""
+Tests for blockmean.
+"""
+import os
+
+import numpy.testing as npt
+import pandas as pd
+import pytest
+from pygmt import blockmean
+from pygmt.datasets import load_sample_bathymetry
+from pygmt.exceptions import GMTInvalidInput
+from pygmt.helpers import GMTTempFile, data_kind
+
+
+def test_blockmean_input_dataframe():
+    """
+    Run blockmean by passing in a pandas.DataFrame as input.
+    """
+    dataframe = load_sample_bathymetry()
+    output = blockmean(table=dataframe, spacing="5m", region=[245, 255, 20, 30])
+    assert isinstance(output, pd.DataFrame)
+    assert all(dataframe.columns == output.columns)
+    assert output.shape == (5849, 3)
+    npt.assert_allclose(output.iloc[0], [245.888877, 29.978707, -384.0])
+
+    return output
+
+
+def test_blockmean_wrong_kind_of_input_table_matrix():
+    """
+    Run blockmean using table input that is not a pandas.DataFrame but still a
+    matrix.
+    """
+    dataframe = load_sample_bathymetry()
+    invalid_table = dataframe.values
+    assert data_kind(invalid_table) == "matrix"
+    with pytest.raises(GMTInvalidInput):
+        blockmean(table=invalid_table, spacing="5m", region=[245, 255, 20, 30])
+
+
+def test_blockmean_wrong_kind_of_input_table_grid():
+    """
+    Run blockmean using table input that is not a pandas.DataFrame or file but
+    a grid.
+    """
+    dataframe = load_sample_bathymetry()
+    invalid_table = dataframe.bathymetry.to_xarray()
+    assert data_kind(invalid_table) == "grid"
+    with pytest.raises(GMTInvalidInput):
+        blockmean(table=invalid_table, spacing="5m", region=[245, 255, 20, 30])
+
+
+def test_blockmean_input_filename():
+    """
+    Run blockmean by passing in an ASCII text file as input.
+    """
+    with GMTTempFile() as tmpfile:
+        output = blockmean(
+            table="@tut_ship.xyz",
+            spacing="5m",
+            region=[245, 255, 20, 30],
+            outfile=tmpfile.name,
+        )
+        assert output is None  # check that output is None since outfile is set
+        assert os.path.exists(path=tmpfile.name)  # check that outfile exists at path
+        output = pd.read_csv(tmpfile.name, sep="\t", header=None)
+        assert output.shape == (5849, 3)
+        npt.assert_allclose(output.iloc[0], [245.888877, 29.978707, -384.0])
+
+    return output
+
+
+def test_blockmean_without_outfile_setting():
+    """
+    Run blockmean by not passing in outfile parameter setting.
+    """
+    with pytest.raises(GMTInvalidInput):
+        blockmean(table="@tut_ship.xyz", spacing="5m", region=[245, 255, 20, 30])
```

Comment:
One thing I've been contemplating is whether to just give people a `pandas.DataFrame` when no argument is passed into the `outfile` parameter. I feel like the current way is too restrictive, why can't people just pass in a filename (str) input to `table` and get a `pandas.DataFrame` directly? If they set an `outfile`, then return `None` by all means, but I don't think we should force `table=filename` inputs to always go into the `outfile`. What do you think?

Reply:
I agree it's too restrictive. The input could be a file, a matrix, a pandas.DataFrame or x/y/z.
The output could be a file or a pandas.DataFrame.

Reply:
I also agree. Do you want me to add that to this PR? I would tend towards implementing it separately, since it will impact both blockmean and blockmedian.

Reply:
I was trying to improve the blockmean wrapper yesterday and realized that it will cause conflicts with this PR. I think we should improve it in a separate PR, either before or after this PR is merged.

Reply:
After seems better to me since this PR is done as far as I can tell and that PR is not open yet. Do you want to work on improving the wrapper? I'm glad to help with adding DataFrame output support, but also fine with leaving it to you in which case I would wait until after you are done to open the PR for wrapping blockmode.

Reply:
I'll leave it to you or @weiji14, as I have never used blockm*.

Reply:
Sounds good. I'll open an issue to keep track of this proposal.